File: src/ShallowWater.cpp
/*!
 * @file ShallowWater.cpp
 * @author Panos Asproulis <p.asproulis@icloud.com>
 * @version 2.1
 *
 * @brief Computes the analytical solution of the shallow water flow over a
 * bump.
 */

#include "ShallowWater.h"

std::tuple<double, double, double, double> ShallowWater::bernoulli_coefficients(
    double q_in, double h_out, double z_in, double z_out, double g)
{
    const double a = 1.0;
    const double b = -(q_in * q_in / (2.0 * g * h_out * h_out) + h_out - (z_in - z_out));
    const double c = 0.0;
    const double d = q_in * q_in / (2.0 * g);

    return std::make_tuple(a, b, c, d);
}

std::vector<std::complex<double>>
ShallowWater::solve_cubic_equation(std::tuple<double, double, double, double> coefficients)
{
    double a = std::get<0>(coefficients);
    double b = std::get<1>(coefficients);
    double c = std::get<2>(coefficients);
    double d = std::get<3>(coefficients);
    //
    // Normalize the coefficients.
    //
    b /= a;
    c /= a;
    d /= a;
    //
    // Calculate the discriminant.
    //
    const double delta0 = b * b - 3.0 * c;
    const double delta1 = 2.0 * b * b * b - 9.0 * b * c + 27.0 * d;

    std::complex<double> C = std::pow(
        (delta1 +
         std::sqrt(std::complex<double>(delta1 * delta1 - 4.0 * delta0 * delta0 * delta0))) /
            2.0,
        1.0 / 3.0);

    const std::complex<double> omega(-0.5,
                                     std::sqrt(3.0) / 2.0); // Cube roots of unity
    //
    // Compute the roots.
    //
    std::vector<std::complex<double>> roots(3);
    for (size_t k = 0; k < 3; ++k)
    {
        roots[k] =
            -1.0 / 3.0 * (b + std::pow(omega, k) * C + delta0 / (std::pow(omega, k) * C));
    }

    return roots;
}

double ShallowWater::find_solution(const std::vector<std::complex<double>>& roots,
                                   double h_near)
{
    //
    // Ignore the roots with imaginary parts and negative
    // real parts.
    //
    std::vector<double> real_roots;
    for (const auto& root : roots)
    {
        if (std::fabs(root.imag()) <= 1.0e-6 && root.real() >= 0.0)
        {
            real_roots.push_back(root.real());
        }
    }
    //
    // Select the root which is closest to the previously
    // computed height of water.
    //
    double solution = 0.0;
    double min_diff = std::numeric_limits<double>::max();

    for (const auto& root : real_roots)
    {
        const double diff = std::fabs(root - h_near);
        if (diff < min_diff)
        {
            solution = root;
            min_diff = diff;
        }
    }

    return solution;
}

File: src/shallow_water.rs
// -------------------------------------------------------------------------------------------------
//
//  bump - Computing the height of water for the case of a shallow water flow over a bump
//
//  Copyright (c) 2025 by Dr. Panos Asproulis (p.asproulis@icloud.com).
//  All Rights Reserved.
//
// -------------------------------------------------------------------------------------------------

//
// Import of external crates
//
use num_complex::Complex64;

///
/// Compute the Bernoulli coefficients i.e. the coefficient of the cubic form
/// of the Bernoulli equation a*h^3 + b*h^2 + c*h + d = 0 for the shallow
/// water flow between an inflow and an outflow location.
///
/// - Arguments:
///   - 'q_in': inflow discharge (m²/s).
///   - 'h_out': outflow height (m).
///   - 'z_in': inflow elevation (m).
///   - 'z_out': outflow elevation (m).
///   - 'g': acceleration due to gravity (m/s²).
///
/// - Returns:
///   - The Bernoulli coefficients a, b, c, d.
///
pub fn bernoulli_coefficients(
    q_in: f64,
    h_out: f64,
    z_in: f64,
    z_out: f64,
    g: f64,
) -> (f64, f64, f64, f64) {
    let a = 1.0;
    let b = -(q_in * q_in / (2.0 * g * h_out * h_out) + h_out - (z_in - z_out));
    let c = 0.0;
    let d = q_in * q_in / (2.0 * g);

    (a, b, c, d)
}

///
/// Find the solution which makes sense for a particular case
/// given the three roots of the cubic equation.
///
/// - Arguments:
///   - 'roots': the roots of the cubic equation.
///   - 'h_near': The height of water at the previously computed location (m).
///
/// - Returns:
///   - The solution which makes sense.
///
pub fn find_solution(roots: &[Complex64], h_near: f64) -> f64 {
    //
    // Ignore the roots with imaginary parts and negative
    // real parts.
    //
    let mut real_roots: Vec<f64> = Vec::new();
    for x in roots.iter() {
        if x.im.abs() <= 1.0e-6 && x.re >= 0.0 {
            real_roots.push(x.re);
        }
    }
    //
    // Select the root which is closest to the previously
    // computed height of water.
    //
    let mut solution: f64 = 0.0;
    let mut min_diff: f64 = f64::MAX;
    for root in real_roots.iter() {
        let diff = (root - h_near).abs();
        if diff < min_diff {
            min_diff = diff;
            solution = *root;
        }
    }

    solution
}
